### ISSUE TITLE
Inconsistent Q when using nb_max_episode_steps

### ISSUE THREAD

**Issue by jurgisp** [keras-rl/keras-rl#298]
When terminating episode early with `nb_max_episode_steps`, the state transition is registered with `terminal=true` flag. This causes DQN to set discounted reward to zero for that state, which is incorrect, because the state might be indistinguishible from any other state. This has the following negative effects:

- Loss doesn't converge to zero, even if the environment is perfectly deterministic and solvable
- Q converges to an incorrect value

For example, running [CartPole example](https://github.com/keras-rl/keras-rl/blob/master/examples/dqn_cartpole.py) with `nb_max_episode_steps=5`:

```
Interval 1 (0 steps performed)
200 episodes - episode_reward: 5.000 [5.000, 5.000] - loss: 0.656 - mean_q: 3.227

Interval 2 (1000 steps performed)
200 episodes - episode_reward: 5.000 [5.000, 5.000] - loss: 2.125 - mean_q: 5.654

Interval 3 (2000 steps performed)
200 episodes - episode_reward: 5.000 [5.000, 5.000] - loss: 2.498 - mean_q: 5.838

Interval 4 (3000 steps performed)
200 episodes - episode_reward: 5.000 [5.000, 5.000] - loss: 2.516 - mean_q: 5.980

...
```

We can see that Q converges to `5.9` and loss stays high. 

The expected behaviour would be for Q to converge to `100.0` (discounted reward with `gamma=0.99` if reward is always `1.0`), and loss should approach zero.

This is an extreme example, but the problem occurs even if max steps is 100 or 500 - the converging properties of Q and loss functions are changed significantly. Conceptually, the problem is that the agent unexpectedly receives a terminal state (with zero Q-value), when judging from the input, it is a perfectly ordinary state with high Q-value.

To address this issue, we should distinguish two flags:

- `terminal`, meaning that it is a terminal state in MDP with zero discounted reward. This should be used [here](https://github.com/keras-rl/keras-rl/blob/master/rl/agents/dqn.py#L268)
- `episode_end`, meaning it is a state after which episode was reset. This should be used [here](https://github.com/keras-rl/keras-rl/blob/master/rl/memory.py#L217)

Then `backward()` function should really take both these parameters, and [max steps reset](https://github.com/keras-rl/keras-rl/blob/master/rl/core.py#L193) should only set `episode_end=True` but not `terminal=True`.

Let me know what you think, if it makes sense, I could try to create a PR.

**Comment by Qinusty** [comment]
From what I can see this is the future rewards being zero'd. This makes sense as from a terminal state there are no future rewards.

This can be seen [here](https://github.com/keras-rl/keras-rl/blob/master/rl/agents/dqn.py#L298) as the q_batch is the predicted future reward from the next state (state1). 

This is my understanding of the code after reviewing it after seeing this issue ticket. Feel free to correct me as I am not 100% familiar with the code base.

**Comment by jurgisp** [comment]
Yes, when `nb_max_episode_steps` is reached, the final state is treated as terminal state, and future reward is set to zero. My suggestion is that this shouldn't be so, because the final state in that case is not actually a terminal state in the environment, but just arbitrary reset by the learning agent.

**Comment by tahaeghtesad** [comment]
I have an environment which doesn't have a terminal state. When I provide `nb_max_episode_steps`, _keras-rl_ treats the state with `step = nb_max_episode_steps` as terminal and disregard future values. As this state may come in other steps, too, _keras-rl_ **should not** set its future reward to zero.
